atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/knowledge/KnownConcludedTransactionsStore.java

```diff
@@ -0,0 +1,166 @@
+            try {
+                keyValueService.checkAndSet(checkAndSetRequest);
+                return;
+            } catch (CheckAndSetException checkAndSetException) {
+                // swallow for retrying
+                log.info(
+                        "Attempt to update the KVS for the timestamp range store failed, possibly because someone "
+                                + "else wrote to this table concurrently.",
+                        checkAndSetException);
+            }
+        }
+        log.warn("Unable to supplement the set of concluded timestamps with a new timestamp range. This may be "
+                + "because the database is momentarily unavailable, or because of particularly high contention.");
+        throw new SafeIllegalStateException("Unable to supplement set of concluded timestamps.");
+    }
+
+    private boolean isRangeContained(Optional<ReadResult> containingRanges, Range<Long> timestampRangeToTest) {
+        return containingRanges
+                .map(ReadResult::timestampRangeSet)
+                .map(ranges -> ranges.encloses(timestampRangeToTest))
+                .orElse(false);
+    }
+
+    private CheckAndSetRequest getCheckAndSetRequest(Optional<ReadResult> oldValue, Range<Long> timestampRangeToAdd) {
+        byte[] serializedTargetSet = serializeTimestampRangeSet(
+                getTargetSet(oldValue.map(ReadResult::timestampRangeSet), timestampRangeToAdd));
+
+        if (oldValue.isEmpty()) {
+            return CheckAndSetRequest.newCell(tableReference, valueCell, serializedTargetSet);
+        }
+        return CheckAndSetRequest.singleCell(
+                tableReference, valueCell, oldValue.get().valueReadFromDatabase(), serializedTargetSet);
+    }
+
+    private byte[] serializeTimestampRangeSet(TimestampRangeSet targetSet) {
+        try {
+            return OBJECT_MAPPER.writeValueAsBytes(targetSet);
+        } catch (JsonProcessingException e) {
+            log.warn("Error serializing timestamp range set", SafeArg.of("targetSet", targetSet), e);
+            throw new RuntimeException(e);
+        }
+    }
+
+    private TimestampRangeSet getTargetSet(Optional<TimestampRangeSet> originalSet, Range<Long> timestampRangeToAdd) {
+        if (originalSet.isEmpty()) {
+            return TimestampRangeSet.singleRange(timestampRangeToAdd);
+        }
+        return originalSet.get().copyAndAdd(timestampRangeToAdd);
+    }
+
+    private Optional<ReadResult> getInternal() {
+        Map<Cell, Value> read = keyValueService.get(tableReference, ImmutableMap.of(valueCell, Long.MAX_VALUE));
+        return read.values().stream().findAny().map(Value::getContents).map(bytes -> ImmutableReadResult.builder()
+                .valueReadFromDatabase(bytes)
+                .build());
+    }
+
+    @org.immutables.value.Value.Immutable
+    interface ReadResult {
+        byte[] valueReadFromDatabase();
+
+        @org.immutables.value.Value.Lazy
```

Comment:
ah nice, I was reading top down and was not happy that this was being called repeatedly